components/director/internal/open_resource_discovery/service.go

```diff
@@ -117,123 +106,66 @@ func (s *Service) SyncORDDocuments(ctx context.Context) error {
-	page, err := s.appSvc.ListGlobal(ctx, pageSize, cursor)
-	if err != nil {
-		return nil, err
-	}
 
-	if len(page.Data) > 0 {
-		applicationIDs := make([]string, 0, page.TotalCount)
-		for _, app := range page.Data {
-			applicationIDs = append(applicationIDs, app.ID)
-		}
+	ctx = persistence.SaveToContext(ctx, tx)
 
-		labels, err := s.labelRepo.ListGlobalByKeyAndObjects(ctx, model.ApplicationLabelableObject, applicationIDs, applicationTypeLabel)
+	if webhook.ObjectType == model.ApplicationTemplateWebhookReference {
+		appTemplateID := webhook.ObjectID
+		apps, err := s.getApplicationsForAppTemplate(ctx, appTemplateID)
 		if err != nil {
-			return nil, err
-		}
-
-		appLabelsMap := make(map[string]interface{}, page.TotalCount)
-		for i := range labels {
-			appLabelsMap[labels[i].ObjectID] = labels[i].Value
+			return err
 		}
-
-		for i := range page.Data {
-			appType, ok := appLabelsMap[page.Data[i].ID].(string)
-			if ok {
-				page.Data[i].Type = appType
+		for _, app := range apps {
+			ctx, err = s.saveTenantToContext(ctx, app.ID)
+			if err != nil {
+				return err
+			}
+			if _, err := s.appSvc.GetForUpdate(ctx, app.ID); err != nil {
+				return errors.Wrapf(err, "error while locking app with id %q for update", app.ID)
+			}
+			if err := s.processWebhookAndDocuments(ctx, tx, webhook, app, globalResourcesOrdIDs); err != nil {
+				return err
 			}
 		}
-	}
-
-	return page, tx.Commit()
-}
-
-func (s *Service) processApp(ctx context.Context, app *model.Application, globalResourcesOrdIDs map[string]bool, appTemplatesWebhooks map[string][]*model.Webhook) error {
-	tx, err := s.transact.Begin()
-	if err != nil {
-		return err
-	}
-
-	defer s.transact.RollbackUnlessCommitted(ctx, tx)
-
-	ctx = persistence.SaveToContext(ctx, tx)
-
-	internalTntID, err := s.tenantSvc.GetLowestOwnerForResource(ctx, resource.Application, app.ID)
-	if err != nil {
-		return err
-	}
-
-	tnt, err := s.tenantSvc.GetTenantByID(ctx, internalTntID)
-	if err != nil {
-		return err
-	}
-
-	ctx = tenant.SaveToContext(ctx, internalTntID, tnt.ExternalTenant)
-
-	if _, err := s.appSvc.GetForUpdate(ctx, app.ID); err != nil {
-		return errors.Wrapf(err, "error while looking app with id %q for update", app.ID)
-	}
-
-	var webhookExecuted bool
-	appWebhooks, err := s.webhookSvc.ListForApplicationWithSelectForUpdate(ctx, app.ID)
-	if err != nil {
-		return errors.Wrapf(err, "error fetching webhooks for app with id %q", app.ID)
-	}
-	if webhookExecuted, err = s.processWebhooksAndDocuments(ctx, tx, appWebhooks, app, globalResourcesOrdIDs); err != nil {
-		return err
-	}
-
-	if app.ApplicationTemplateID != nil && !webhookExecuted {
-		appTemplateWebhooks := appTemplatesWebhooks[*app.ApplicationTemplateID]
-		if _, err := s.processWebhooksAndDocuments(ctx, tx, appTemplateWebhooks, app, globalResourcesOrdIDs); err != nil {
+	} else if webhook.ObjectType == model.ApplicationWebhookReference {
+		appID := webhook.ObjectID
+		ctx, err := s.saveTenantToContext(ctx, appID)
```

Comment:
Could you rename this function saveTenantToContext to something else as the name is a bit misleading? At first, I thought that we save the appID in the tnt context but when I saw the implementation it was a bit different from what I thought.

Reply:
is `getLowestOwnerForApp` ok?

Reply:
maybe saveLowestOwnerForAppToContext? How does that sound to you?

Reply:
sounds good